drake/core/Gradient.h

```diff
@@ -18,55 +18,153 @@ namespace Drake {
-    constantTaylorVecXd(val, auto_diff_vec);
-    return auto_diff_vec;
+  /** \brief determine the total size at compile time of a number of arguments based on their SizeAtCompileTime static members
+   */
+  template<typename ...Args>
+  constexpr int totalSizeAtCompileTime() {
+    return internal::TotalSizeAtCompileTime<Args...>::eval();
   }
 
+  /** \brief determine the total size at runtime of a number of arguments using their size() methods (base case).
+   */
+  constexpr Eigen::DenseIndex totalSizeAtRunTime() {
+    return 0;
+  }
 
-  /** initTaylorVecXd
-   * @brief initializes the vector with x=value and dx=gradient
+  /** \brief determine the total size at runtime of a number of arguments using their size() methods (recursive)
    */
-  template<typename Derived, typename DerivedGradient>
-  void initTaylorVecXd(const Eigen::MatrixBase<Derived>& val, const Eigen::MatrixBase<DerivedGradient>& gradient, TaylorVecXd& auto_diff_vector)
-  {
-    typedef typename Eigen::MatrixBase<DerivedGradient>::Index Index;
-    auto_diff_vector.resize(gradient.size(),1);
-    auto nx = gradient.cols();
-    for (Index row = 0; row < auto_diff_vector.size(); row++) {
-      auto_diff_vector(row).value() = val(row);
-      auto_diff_vector(row).derivatives().resize(nx,1);
-      auto_diff_vector(row).derivatives() = gradient.row(row).transpose();
-    }
+  template<typename Head, typename ...Tail>
+  Eigen::DenseIndex totalSizeAtRunTime(const Eigen::MatrixBase<Head> &head, const Tail &... tail) {
+    return head.size() + totalSizeAtRunTime(tail...);
+  }
+  // TODO end: move to different location?
+
+  namespace internal {
+    /** \brief helper for initializeAutoDiffArgs function (recursive)
+     */
+    template<size_t Index>
+    struct InitializeAutoDiffArgsHelper {
+      template<typename ...ValueTypes, typename ...AutoDiffTypes>
+      static void run(const std::tuple<ValueTypes...> &values, std::tuple<AutoDiffTypes...> &auto_diffs, Eigen::DenseIndex num_derivatives, Eigen::DenseIndex deriv_num_start) {
+        constexpr size_t tuple_index = sizeof...(AutoDiffTypes) - Index;
+        const auto& value = std::get<tuple_index>(values);
+        auto& auto_diff = std::get<tuple_index>(auto_diffs);
+        auto_diff.resize(value.rows(), value.cols());
+        initializeAutoDiff(value, auto_diff, num_derivatives, deriv_num_start);
+        InitializeAutoDiffArgsHelper<Index - 1>::run(values, auto_diffs, num_derivatives, deriv_num_start + value.size());
+      }
+    };
+
+    /** \brief helper for initializeAutoDiffArgs function (base case)
+     */
+    template<>
+    struct InitializeAutoDiffArgsHelper<0> {
+      template<typename ...ValueTypes, typename ...AutoDiffTypes>
+      static void run(const std::tuple<ValueTypes...> &values, const std::tuple<AutoDiffTypes...> &auto_diffs, Eigen::DenseIndex num_derivatives, Eigen::DenseIndex deriv_num_start) {
+        // empty
+      }
+    };
+  }
+
+  template<typename ...Args>
```

Comment:
absolutely imperative that you document this one.  i have absolutely no idea what it's doing from just looking quickly at the code.